stdlib/either.mli

```diff
@@ -0,0 +1,88 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*         Gabriel Scherer, projet Parsial, INRIA Saclay                  *)
+(*                                                                        *)
+(*   Copyright 2019 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Either type.
+
+    Either is the simplest and most generic sum type. It is isomorphic
+    to the [result] type; it should be used when there is no implied
+    success/failure semantics in the choice of one of the constructors.
+
+    @since 4.11 *)
+
+(* Unlike [result], no [either] type is made available in Stdlib,
+   one needs to access [Either.t] explicitly:
+
+   - This type is less common in typical OCaml codebases,
+     which prefer domain-specific variant types whose constructors
+     carry more meaning.
+   - Adding this to Stdlib would raise warnings in existing codebases
+     that already use a constructor named Left or Right:
+     + when opening a module that exports such a name,
+       warning 45 is raised
+     + adding a second constructor of the same name in scope kicks
+       in the disambiguation mechanisms, and warning 41 may now
+       be raised by existing code.
+
+   If the use becomes more common in the future we can always
+   revisit this choice.
+*)
+
+type ('a, 'b) t = Left of 'a | Right of 'b (**)
+(** A value of [('a, 'b) Either.t] contains either a value of ['a]  or a value
+    of ['b] *)
+
+val left : 'a -> ('a, 'b) t
+(** [left v] is [Left v]. *)
+
+val right : 'b -> ('a, 'b) t
+(** [right v] is [Right v]. *)
+
+val get_left : ('a, 'b) t -> 'a option
+(** [get_left (Left v)] is [Some v], [get_left (Right _)] is [None] *)
+
+val get_right : ('a, 'b) t -> 'b option
+(** [get_right (Right v)] is [Some v], [get_right (Left _)] is [None] *)
+
+val swap : ('a, 'b) t -> ('b, 'a) t
+(** [swap (Left v)] is [Right v], and [swap (Right v)] is [Left v]. *)
```

Comment:
:roll_eyes:

Reply:
No strong opinion, but we have `Fun.flip` with similar structure, maybe this could be `flip`.

Reply:
I changed to use `flip`.

The use-case for this function is when someone choose a particular ordering between Left and Right for an API (for example they provide you with a function from `('a, unit) Either.t` to `'a option`, but not one for `(unit, 'a) Either.t`), and your data naturally comes in the other order.